org.scala-ide.sdt.debug/src/org/scalaide/debug/internal/model/ScalaDebugCache.scala

```diff
@@ -244,15 +265,37 @@ protected[debug] class ScalaDebugCacheActor(debugCache: ScalaDebugCache, debugTa
       classLoaded(e)
       reply(false)
     case LoadedNestedTypes(outerTypeName) =>
-      reply(LoadedNestedTypesAnswer(getLoadedNestedTypes(outerTypeName)))
+      val replyTo = sender
+      become {
+        case LoadedNestedTypesUpdate(outerTypeName, cache) =>
+          updateCacheWith((outerTypeName, cache))
+          replyTo ! LoadedNestedTypesAnswer(cache.types)
+          unbecome()
+      }
+      findCache(outerTypeName).onSuccess {
+        case (typeName, cache) =>
+          ScalaDebugCacheActor.this ! LoadedNestedTypesUpdate(typeName, cache)
+      }
     case AddClassPrepareEventListener(actor, outerTypeName) =>
-      addClassPreparedEventListener(actor, outerTypeName)
-      reply(true)
+      val replyTo = sender
+      become {
```

Comment:
there is a problem that unfortunately actor's thread can be killed in any moment when info about types is retrieved and cached and then the sync lock is not released (I didn't find the way how to release lock when scala actor dies). So I had to push out this JDI operation to Future (outside actor's thread). This is connected also to your second question about 'findNestedTypeInJdi'. This method because of reason about needs to be called in Future thread. I hope I'm clear.